rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+<!-- mdlint off(HEADERS_TOO_MANY_H1) -->
+
+# Standardized TFX Inputs
+
+Status        | Proposed
+:------------ | :------------------------------------------------------------
+**RFC #**     | [162](https://github.com/tensorflow/community/pull/162)
+**Author(s)** | Zhuo Peng (zhuo@google.com), Kester Tong (kestert@google.com)
+**Sponsor**   | Konstantinos Katsiapis (katsiapis@google.com)
+**Updated**   | 2019-10-03
+
+# Objective
+
+*   To define a common in-memory data representation that:
+    *   is powerful enough to encode the following logical training data format:
+        flat
+        ([`tf.Example`](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/core/example/example.proto#L88)),
+        sequence
+        ([`tf.SequenceExample`](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/core/example/example.proto#L298))
+        or structured data (e.g.
+        [Protocol Buffers](https://developers.google.com/protocol-buffers) or
+        [Apache Avro](https://avro.apache.org/)).
+    *   all TFX components can understand and can support their own unique use
+        cases with.
+*   To define an I/O abstraction layer that produces the above in-memory
+    representation from supported physical storage formats, while hiding TFX’s
+    choice of such storage formats from TFX users.
+*   To define a bridge from the above in-memory representation to TF feedables
+    (i.e. Tensors and certain
+    [CompositeTensors](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/python/framework/composite_tensor.py#L1)).
+
+# Motivation
+
+## Fragmented in-memory data representations across TFX
+
+A TFX component may use its input data (data generated by ExampleGen) in two
+ways:
+
+*   it may need to understand the data and conduct analysis. Usually this
+    happens in [Apache Beam](https://beam.apache.org/), and does not involve a
+    TF Model. For example:
+    *   [TFDV](https://github.com/tensorflow/data-validation) and
+        [TFT](https://github.com/tensorflow/transform) compute some statistics
+        over the data.
+    *   [TFMA](https://github.com/tensorflow/model-analysis) may slice the
+        dataset by certain columns in the data.
+*   it may feed the data to TensorFlow. Note that feeding TF alone may not
+    require understanding the data. For example, TFMA may feed a TF model with
+    serialized tf.Example, which may be the raw form of the data.
+
+Currently, each TFX component has its own in-memory data representation to cover
+the two use cases by a different approach:
+
+|    | TFDV | TFT  | TFMA | BulkInference |
```

Comment:
TensorFlow model (Trainer) can consume  any form of in-memory data format as input, which is convertd to tensors, in input_fn(), typically by tf.data.Dataset for training, in-memory for serving.

In theory, there is nothing precluding me from implementing a Custom ExampleGen that produces a CSV or JSON, and a Trainer that consumes CSV or JSON as an input, so long as my input_fn does the right thing to consume CSV or JSON.

I would like some discussion on anticipated future direction after this proposal. With introduction of TFXIO, would it move the system to the direction to tighten such looseness (of in-memory and on-disk format for TensorFlow model itself), or, would it be orthogonal so that Trainer (model) continues to define own ways to define input format as it needs.

If it is in the direction to tighten it, what is anticipated impact on custom components such as [1, 2], which can choose to define own format conversion of Examples (such as CSV, JSON), as depicted in [3].

[1]  https://www.tensorflow.org/tfx/guide/custom_component
[2]  https://github.com/tensorflow/community/pull/146
[3]  https://www.tensorflow.org/tfx/guide#artifacts

Reply:
The direction question is worth discussing in the design review.

Technically though, I could imagine a custom component adopting TFXIO to gain understanding of the data. I could also imagine that a separate set of API to allow custom ExampleGens to write data, into something opaque, so the rest of the pipeline still uses TFXIO to gain the access. That being said, it's still possible to not adopt TFXIO, and not  (or not let native TFX components to) understand the data, or establish some agreement among certain custom components.